### PR TITLE
Fix scroll-jumps when jumping to a permalink

### PR DESCRIPTION
**Pull request by ara4n**
Having turned on scroll debugging, it looks like
https://github.com/vector-im/element-web/issues/23393 is caused by scroll events firing while forward-filling is happening, which causes `ScrollPanel`'s `onScroll()` to lock the scroll position to the bottom of the page via `saveScrollState()`, rather than honour the permalink's scroll offset.

Either we could make `saveScrollState()` honour permalinks, or just ignore scroll events which fire while filling - as `checkFillState()` will go and apply the scroll-lock anyway once the filling is complete.

This fix attempts the latter; i've manually tested a bunch (~20 times) and it /seems/ to have fixed it? It was pretty hard to repro in the first place, though.

Hopefully fixes https://github.com/vector-im/element-web/issues/23393 and [chronic bug](https://docs.google.com/spreadsheets/d/174m6bfl-v5y3BELQ-VNpyBLrLDLfJzW7-5R8chxESY0/edit#gid=0) 89.

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix scroll-jumps when jumping to a permalink ([\#9340](https://github.com/matrix-org/matrix-react-sdk/pull/9340)). Fixes vector-im/element-web#23393.

### CODE REVIEW

**Comment by ara4n** [comment]
(this may require review from @richvdh or @bwindels or whoever else knows the darker secrets of ScrollPanel)

**Comment by ara4n** [comment]
btw, there aren't tests on this because I don't think we've figured out how to test scrolling at all yet - which might explain why it's such a mess. one option could be to have a mode which consciously delays most scroll activity (backfilling, etc) in order to make it easier to reason about and test.

I've tested this manually to check whether it breaks scroll behaviour when scrollbacking, but it seems to be okay (surprisingly, given, you'd expect scroll events to fire while backpagination is happening, which get discarded, causing jumps. but it might actually have /fixed/ scrolljumps while backpaginating perhaps?)

**Review by SimonBrandner — APPROVED** [review]
LGTM

**Review by justjanne — APPROVED** [review]

File: src/components/structures/ScrollPanel.tsx
Comment (MadLittleMods) [inline src/components/structures/ScrollPanel.tsx:227]: Let's ship this 🙇 

Merged `develop` so we can ensure it's still good to go test wise even though there are no scroll tests.

Comment (MadLittleMods) [reply src/components/structures/ScrollPanel.tsx:227]: [TypeScript strict CI job is failing](https://github.com/matrix-org/matrix-react-sdk/actions/runs/3481716623/jobs/5823161509) but it's complaining about an adjacent unrelated line of code and that job is not required to pass.

Cypress flake that passed after the second run ✅

Everything looks okay to merge ⏩

**Comment by turt2live** [comment]
from oob: we're not sure if this actually fixes an issue at the moment, and needs a regression test to do so. Writing a test for this will be challenging until https://github.com/vector-im/element-web/issues/23531 is complete. This PR might also conflict with https://github.com/vector-im/element-web/issues/23539

Marking as blocked until one of those two issues is closed, at which point we can re-evaluate this.

File: src/components/structures/ScrollPanel.tsx
Comment (MadLittleMods) [inline src/components/structures/ScrollPanel.tsx:237]: > we're not sure if this actually fixes an issue at the moment, and needs a regression test to do so.

@ara4n thinks the problem might still be reproducible even with this fix but do these changes at least make sense?

We should merge in that case anyway

**Comment by Johennes** [comment]
It's true that we don't have the bandwidth to devote to fixing scroll jumpiness properly (which would also include tests). Meanwhile though I don't think we necessarily need to block all contributions that aim at improving the situation. I think this should go through normal review again (because the original one was a year ago) and have a few folks test it. I've resolved the conflicts and requested another review from matrix-org/element-web.

**Comment by claell** [comment]
Hm, testing the Netlify build, I don't really see any things which got worse. However, also no real improvement. Didn't really have a permalink to jump to, but used the unread message button to jump up. Did only test once, maybe there even was some improvement to that which I didn't notice? At least, it still seems broken.

grafik

By the way, with that build (compared to development.element.io), this icon does only show, if I scroll up. Otherwise, it's hidden. Not sure whether this is introduced by this PR, whether this is wanted, or another bug. Just wanted to mention it.

**Comment by dbkr** [comment]
This seems to be causing a read receipts test to not request a read marker which is a distinctly non-obvious and subtle failure mode...

**Comment by dbkr** [comment]
Closing now for as the test failure looks like it could be legitimate and doesn't look trivial to fix. Hopefully we can revisit this later as part of scroll jump investigation.

**Pull request closed without merging**
